### PR TITLE
[C5][environment][sso] Replace uiService query parameter with a configuration

### PR DESCRIPTION
**Pull request by RenukaFernando**
### Proposed changes in this pull request
- Use the first host in the `allowedOrigins` list as UI Service.
- If the first host is empty use the `wso2.carbon.apimgt.application: apimBaseUrl` as UI Service
- Configurations are changed as below.

From:
```yaml
wso2.carbon.apimgt:
    # Environment Configurations
  environmentConfigurations:
      # list of web clients (eg: 127.0.0.1:9292) to allow make requests to current environment
      # (use '*' to allow any web client)
    allowedHosts:
    - 10.100.4.247:9292
    - localhost:9292
      # current environment's label from the list of environments
    environmentLabel: Development
```
To:
```yaml
wso2.carbon.apimgt:
    # Environment Configurations
  environmentConfigurations:
      # current environment's label from the list of environments
    environmentLabel: Default
      # list of web clients (eg: 127.0.0.1:9292) to allow make requeststo current environment
      # (use '*' to allow any web client)
      # the first host is used as UI-Service
    allowedHosts:
    - ''
```

- Remove the `uiService` query parameter from callback URL.
- Fix https://github.com/wso2/product-apim/issues/2205

### CODE REVIEW

**Comment by pubudu538** [comment]
I guess allowedHosts is for CORs handling. Why are we removing that? Can you explain the use case here?

**Comment by RenukaFernando** [comment]
`uiService` is used for both handling CORS and redirecting to UI Service when SSO enabled.
- `allowedHosts` is also a list of allowed UI Services. Since it's one UI Service it doesn't need a list.
- Passing `uiService` as a query parameter is a problem.
  - It is also possible to introduce a new configuration for `uiService` with keeping the `allowedHosts` configuration. (Seperate configurations for above two functionalities).

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.authenticator/src/main/java/org/wso2/carbon/apimgt/rest/api/authenticator/AuthenticatorService.java [hunk 1]
```diff
@@ -155,8 +152,7 @@ public AccessTokenInfo getTokens(String appName, String grantType,
                 // Access token for authorization code grant type
                 APIMAppConfigurations appConfigs = ServiceReferenceHolder.getInstance()
                         .getAPIMAppConfiguration();
-                String callBackURL = appConfigs.getApimBaseUrl() + AuthenticatorConstants.AUTHORIZATION_CODE_CALLBACK_URL + appName
-                        + "?uiService=" + uiServiceURL;
+                String callBackURL = appConfigs.getApimBaseUrl() + AuthenticatorConstants.AUTHORIZATION_CODE_CALLBACK_URL + appName;
```
Comment (harsha89) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api.authenticator/src/main/java/org/wso2/carbon/apimgt/rest/api/authenticator/AuthenticatorService.java]: better to have debug log with the callbackurl

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.authenticator/src/main/java/org/wso2/carbon/apimgt/rest/api/authenticator/AuthenticatorAPI.java [hunk 2]
```diff
@@ -319,6 +318,16 @@ public Response callback(@Context Request request, @PathParam("appName") String
                     log.debug("Set cookies for " + appName + " application.");
                 }
                 // Redirect to the store/apis page (redirect URL)
+                String uiServiceUrl;
+                //The first host in the list "allowedHosts" is the host of UI-Service
+                String uiServiceHost = APIMConfigurationService.getInstance().getApimConfigurations()
+                        .getEnvironmentConfigurations().getAllowedHosts().get(0);
+                if (uiServiceHost.isEmpty()) {
```
Comment (harsha89) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api.authenticator/src/main/java/org/wso2/carbon/apimgt/rest/api/authenticator/AuthenticatorAPI.java]: Use stringutils.isEmpty here. If null value returned there will be null pointer exception throws

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.authenticator/src/main/java/org/wso2/carbon/apimgt/rest/api/authenticator/AuthenticatorAPI.java [hunk 3]
```diff
@@ -319,6 +318,16 @@ public Response callback(@Context Request request, @PathParam("appName") String
...
+                    uiServiceUrl = ServiceReferenceHolder.getInstance().getAPIMAppConfiguration().getApimBaseUrl();
+                } else {
+                    uiServiceUrl = "https://" + uiServiceHost + "/";
```
Comment (pubudu538) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api.authenticator/src/main/java/org/wso2/carbon/apimgt/rest/api/authenticator/AuthenticatorAPI.java]: Why are we not giving the complete URL in allow hosts with the protocol?

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.authenticator/src/main/java/org/wso2/carbon/apimgt/rest/api/authenticator/AuthenticatorAPI.java [hunk 4]
```diff
@@ -319,6 +318,16 @@ public Response callback(@Context Request request, @PathParam("appName") String
                     log.debug("Set cookies for " + appName + " application.");
                 }
                 // Redirect to the store/apis page (redirect URL)
+                String uiServiceUrl;
+                //The first host in the list "allowedHosts" is the host of UI-Service
+                String uiServiceHost = APIMConfigurationService.getInstance().getApimConfigurations()
+                        .getEnvironmentConfigurations().getAllowedHosts().get(0);
+                if (uiServiceHost.isEmpty()) {
+                    uiServiceUrl = ServiceReferenceHolder.getInstance().getAPIMAppConfiguration().getApimBaseUrl();
```
Comment (harsha89) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api.authenticator/src/main/java/org/wso2/carbon/apimgt/rest/api/authenticator/AuthenticatorAPI.java]: So why we need host and base url both? When host is provided base url is ommitted even it has specified. We need to be consistent and use single url. Looking at two places will cause issues. Let's add a debug log as well.

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.authenticator/src/main/java/org/wso2/carbon/apimgt/rest/api/authenticator/AuthenticatorAPI.java (hunk 2 above)
Comment (RenukaFernando) [reply components/apimgt/org.wso2.carbon.apimgt.rest.api.authenticator/src/main/java/org/wso2/carbon/apimgt/rest/api/authenticator/AuthenticatorAPI.java]: Done with https://github.com/wso2/carbon-apimgt/pull/4896/commits/a887a18fff7578d20969d7146e46329a165ae57e

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.authenticator/src/main/java/org/wso2/carbon/apimgt/rest/api/authenticator/AuthenticatorAPI.java (hunk 4 above)
Comment (RenukaFernando) [reply components/apimgt/org.wso2.carbon.apimgt.rest.api.authenticator/src/main/java/org/wso2/carbon/apimgt/rest/api/authenticator/AuthenticatorAPI.java]: Done with https://github.com/wso2/carbon-apimgt/pull/4896/commits/131693463cb521f3aa19ae56aab16f2c65550c2c

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.authenticator/src/main/java/org/wso2/carbon/apimgt/rest/api/authenticator/AuthenticatorService.java (hunk 1 above)
Comment (RenukaFernando) [reply components/apimgt/org.wso2.carbon.apimgt.rest.api.authenticator/src/main/java/org/wso2/carbon/apimgt/rest/api/authenticator/AuthenticatorService.java]: Done with https://github.com/wso2/carbon-apimgt/pull/4896/commits/131693463cb521f3aa19ae56aab16f2c65550c2c

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.authenticator/src/main/java/org/wso2/carbon/apimgt/rest/api/authenticator/AuthenticatorAPI.java [hunk 5]
```diff
@@ -319,6 +322,27 @@ public Response callback(@Context Request request, @PathParam("appName") String
...
+                    }
+                    uiServiceUrl = ServiceReferenceHolder.getInstance().getAPIMAppConfiguration().getApimBaseUrl();
+                } else {
+                    if (log.isDebugEnabled()) {
+                        log.debug("The first string in the list " +
+                                "'wso2.carbon.apimgt:environmentConfigurations:allowedHosts' configuration" +
+                                " is not empty. value: " + uiServiceHost);
+                    }
+                    uiServiceUrl = HTTPS_PROTOCOL + PROTOCOL_SEPERATOR + uiServiceHost + URL_PATH_SEPERATOR;
+                    log.info("UI Service: " + uiServiceUrl);
```
Comment (harsha89) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api.authenticator/src/main/java/org/wso2/carbon/apimgt/rest/api/authenticator/AuthenticatorAPI.java]: use  log.info("UI Service {}: ", uiServiceUrl); of logging style

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.authenticator/src/main/java/org/wso2/carbon/apimgt/rest/api/authenticator/AuthenticatorAPI.java [hunk 6]
```diff
@@ -58,6 +60,10 @@
 import javax.ws.rs.core.NewCookie;
 import javax.ws.rs.core.Response;
 
+import static org.wso2.carbon.apimgt.rest.api.authenticator.constants.AuthenticatorConstants.HTTPS_PROTOCOL;
```
Comment (harsha89) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api.authenticator/src/main/java/org/wso2/carbon/apimgt/rest/api/authenticator/AuthenticatorAPI.java]: any reason for use this kind of constant imports?

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.authenticator/src/main/java/org/wso2/carbon/apimgt/rest/api/authenticator/AuthenticatorAPI.java (hunk 5 above)
Comment (RenukaFernando) [reply components/apimgt/org.wso2.carbon.apimgt.rest.api.authenticator/src/main/java/org/wso2/carbon/apimgt/rest/api/authenticator/AuthenticatorAPI.java]: Fixed with https://github.com/wso2/carbon-apimgt/pull/4896/commits/7f72562a3f2a9dc2595f82daf9f6b95e0fae92e9

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.authenticator/src/main/java/org/wso2/carbon/apimgt/rest/api/authenticator/AuthenticatorAPI.java (hunk 6 above)
Comment (RenukaFernando) [reply components/apimgt/org.wso2.carbon.apimgt.rest.api.authenticator/src/main/java/org/wso2/carbon/apimgt/rest/api/authenticator/AuthenticatorAPI.java]: Fixed with https://github.com/wso2/carbon-apimgt/pull/4896/commits/7f72562a3f2a9dc2595f82daf9f6b95e0fae92e9

**Pull request merged**